src/hashes/blake2s.c

```diff
@@ -289,59 +310,54 @@ int blake2s_process(hash_state *md, const unsigned char *in, unsigned long inlen
       return CRYPT_INVALID_ARG;
    }
 
-   while (inlen > 0) {
-      ulong32 left = md->blake2s.curlen;
-      ulong32 fill = 2 * BLAKE2S_BLOCKBYTES - left;
-
+   if (inlen > 0) {
+      unsigned long left = md->blake2s.curlen;
+      unsigned long fill = BLAKE2S_BLOCKBYTES - left;
       if (inlen > fill) {
-         XMEMCPY(md->blake2s.buf + left, in, fill);
-         md->blake2s.curlen += fill;
+         md->blake2s.curlen = 0;
+         XMEMCPY(md->blake2s.buf + left, in, fill); /* Fill buffer */
          blake2s_increment_counter(md, BLAKE2S_BLOCKBYTES);
-         blake2s_compress(md, md->blake2s.buf);
-         XMEMCPY(md->blake2s.buf, md->blake2s.buf + BLAKE2S_BLOCKBYTES, BLAKE2S_BLOCKBYTES);
-         md->blake2s.curlen -= BLAKE2S_BLOCKBYTES;
+         blake2s_compress(md, md->blake2s.buf); /* Compress */
          in += fill;
          inlen -= fill;
-      } else /* inlen <= fill */
-      {
-         XMEMCPY(md->blake2s.buf + left, in, inlen);
-         md->blake2s.curlen += (ulong32)inlen; /* Be lazy, do not compress */
-         in += inlen;
-         inlen -= inlen;
+         while (inlen > BLAKE2S_BLOCKBYTES) {
+            blake2s_increment_counter(md, BLAKE2S_BLOCKBYTES);
+            blake2s_compress(md, in);
+            in += BLAKE2S_BLOCKBYTES;
+            inlen -= BLAKE2S_BLOCKBYTES;
+         }
       }
+      XMEMCPY(md->blake2s.buf + md->blake2s.curlen, in, inlen);
+      md->blake2s.curlen += inlen;
    }
-
    return CRYPT_OK;
 }
 
 int blake2s_done(hash_state *md, unsigned char *out)
 {
-   unsigned char buffer[BLAKE2S_OUTBYTES];
+   unsigned char buffer[BLAKE2S_OUTBYTES] = { 0 };
    unsigned long i;
 
    LTC_ARGCHK(md != NULL);
    LTC_ARGCHK(out != NULL);
 
    /* if(md->blake2s.outlen != outlen) return CRYPT_INVALID_ARG; */
 
-   if (md->blake2s.curlen > BLAKE2S_BLOCKBYTES) {
-      blake2s_increment_counter(md, BLAKE2S_BLOCKBYTES);
-      blake2s_compress(md, md->blake2s.buf);
-      md->blake2s.curlen -= BLAKE2S_BLOCKBYTES;
-      XMEMCPY(md->blake2s.buf, md->blake2s.buf + BLAKE2S_BLOCKBYTES, md->blake2s.curlen);
-   }
+   if (blake2s_is_lastblock(md))
+      return CRYPT_ERROR;
 
-   blake2s_increment_counter(md, (ulong32)md->blake2s.curlen);
+   blake2s_increment_counter(md, md->blake2s.curlen);
    blake2s_set_lastblock(md);
-   XMEMSET(md->blake2s.buf + md->blake2s.curlen, 0, 2 * BLAKE2S_BLOCKBYTES - md->blake2s.curlen); /* Padding */
+   XMEMSET(md->blake2s.buf + md->blake2s.curlen, 0, BLAKE2S_BLOCKBYTES - md->blake2s.curlen); /* Padding */
    blake2s_compress(md, md->blake2s.buf);
 
    for (i = 0; i < 8; ++i) /* Output full hash to temp buffer */
-      STORE32L(md->blake2s.h[i], buffer + sizeof(md->blake2s.h[i]) * i);
+      STORE32L(md->blake2s.h[i], buffer + i * 4);
 
    XMEMCPY(out, buffer, md->blake2s.outlen);
 #ifdef LTC_CLEAN_STACK
-    zeromem(md, sizeof(hash_state));
+   zeromem(buffer, sizeof(buffer));
+   zeromem(md, sizeof(hash_state));
```

Comment:
~~shouldn't we better remove the zero'ing of `md` for both `_done` functions? or is the hash descriptor unusable anyways after having called `done` on it?~~ I just realized that's according to all the other hash implementations...